src/dep_kinds_filtering.rs

```diff
@@ -0,0 +1,262 @@
+use crate::{Args, VendorFilter};
+use anyhow::{Context, Result};
+use camino::Utf8Path;
+use clap::{builder::PossibleValue, ValueEnum};
+use serde::{Deserialize, Serialize};
+use std::{
+    borrow::Cow,
+    collections::{HashMap, HashSet},
+};
+
+/// Kinds of dependencies that shall be included.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+pub enum DepKinds {
+    All,
+    Normal,
+    Build,
+    Dev,
+    NoNormal,
+    NoBuild,
+    NoDev,
+}
+
+impl ValueEnum for DepKinds {
+    fn value_variants<'a>() -> &'a [Self] {
+        &[
+            Self::All,
+            Self::Normal,
+            Self::Build,
+            Self::Dev,
+            Self::NoNormal,
+            Self::NoBuild,
+            Self::NoDev,
+        ]
+    }
+
+    fn to_possible_value(&self) -> Option<PossibleValue> {
+        Some(match self {
+            Self::All => PossibleValue::new("all"),
+            Self::Normal => PossibleValue::new("normal"),
+            Self::Build => PossibleValue::new("build"),
+            Self::Dev => PossibleValue::new("dev"),
+            Self::NoNormal => PossibleValue::new("no-normal"),
+            Self::NoBuild => PossibleValue::new("no-build"),
+            Self::NoDev => PossibleValue::new("no-dev"),
+        })
+    }
+}
+
+impl std::fmt::Display for DepKinds {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.to_possible_value()
+            .expect("No values are skipped")
+            .get_name()
+            .fmt(f)
+    }
+}
+
+/// Filter out unwanted dependency kinds.
+///
+/// Replicates logic from add_packages_for_platform() but uses cargo tree
+/// because cargo metadata does not implement dependency kinds filtering.
+/// Ref: <https://github.com/rust-lang/cargo/issues/10718>
+/// Cargo tree is NOT intended for automatic processing so this function
+/// explicitly does not replace the add_packages_for_platform() entirely.
+pub(crate) fn filter_dep_kinds(
+    args: &Args,
+    config: &VendorFilter,
+    packages: &mut HashMap<cargo_metadata::PackageId, &cargo_metadata::Package>,
+    platform: Option<&str>,
+) -> Result<()> {
+    // exit early when no dependency kinds filtering is requested
+    match config.keep_dep_kinds {
+        None | Some(DepKinds::All) => return Ok(()),
+        Some(_) => (),
+    };
+
+    let required_packages = get_required_packages(
+        &args.get_all_manifest_paths(),
+        args.offline,
+        config,
+        platform,
+    )?;
+
+    packages.retain(|_, package| {
+        required_packages.contains(&(
+            Cow::Borrowed(&package.name),
+            Cow::Borrowed(&package.version),
+        ))
+    });
+    Ok(())
+}
+
+/// Returns the set of required packages to satisfy filters specified in config
+fn get_required_packages<'a>(
+    manifest_paths: &Vec<Option<&Utf8Path>>,
```

Comment:
Very minor nit but generally slices are preferred over `&Vec`, i.e.
```suggestion
    manifest_paths: &[Option<&Utf8Path>],
```